### DISCUSSION TITLE
工具无法正常调用

### DISCUSSION THREAD

**Discussion by shenmadouyaowen** [zai-org/ChatGLM3#335] [Bad Case]
### System Info / 系統信息

ubuntu22.04

### Who can help? / 谁可以帮助到您？

composite_dem @abmfy

### Information / 问题信息

- [X] The official example scripts / 官方的示例脚本
- [X] My own modified scripts / 我自己修改的脚本和任务

### Reproduction / 复现过程

进行多轮对话时无法继续使用工具
![image](https://github.com/THUDM/ChatGLM3/assets/12426724/60bc19d2-ea5d-4de5-9f54-b0d9aa7c85d4)

### Expected behavior / 期待表现

在多轮对话中可以继续使用tool

**Comment by GavinZhou0125** [comment]
我也有同样的问题，即使他第二次模型返回了tool_call 也会出现Failed to parse tool call的现象

**Comment by CNUIGB** [comment]
我的都不会调用工具
![微信截图_20231115115525](https://github.com/THUDM/ChatGLM3/assets/103934790/78d185d6-db7f-4390-ad06-57f8886af9cd)

**Comment by shenmadouyaowen** [comment]
> 我的都不会调用工具 ![微信截图_20231115115525](https://user-images.githubusercontent.com/103934790/283044583-78d185d6-db7f-4390-ad06-57f8886af9cd.png)

只有chatglm3-6b才能使用工具

**Comment by CNUIGB** [comment]
> > 我的都不会调用工具 ![微信截图_20231115115525](https://user-images.githubusercontent.com/103934790/283044583-78d185d6-db7f-4390-ad06-57f8886af9cd.png)
> 
> 只有chatglm3-6b才能使用工具

我的是chatglm3-6b呀

**Comment by shenmadouyaowen** [comment]
> > > 我的都不会调用工具 ![微信截图_20231115115525](https://user-images.githubusercontent.com/103934790/283044583-78d185d6-db7f-4390-ad06-57f8886af9cd.png)
> > 
> > 
> > 只有chatglm3-6b才能使用工具
> 
> 我的是chatglm3-6b呀

看一下后台终端记录

**Comment by CNUIGB** [comment]
> > > > 我的都不会调用工具 ![微信截图_20231115115525](https://user-images.githubusercontent.com/103934790/283044583-78d185d6-db7f-4390-ad06-57f8886af9cd.png)
> > > 
> > > 
> > > 只有chatglm3-6b才能使用工具
> > 
> > 
> > 我的是chatglm3-6b呀
> 
> 看一下后台终端记录

![微信截图_20231115160531](https://github.com/THUDM/ChatGLM3/assets/103934790/c2f7fd0b-3b64-4ada-bf89-749713ae439b)

**Comment by CNUIGB** [comment]
> > > > > 我的都不会调用工具 ![微信截图_20231115115525](https://user-images.githubusercontent.com/103934790/283044583-78d185d6-db7f-4390-ad06-57f8886af9cd.png)
> > > > 
> > > > 
> > > > 只有chatglm3-6b才能使用工具
> > > 
> > > 
> > > 我的是chatglm3-6b呀
> > 
> > 
> > 看一下后台终端记录
> 
> ![微信截图_20231115160531](https://user-images.githubusercontent.com/103934790/283045981-c2f7fd0b-3b64-4ada-bf89-749713ae439b.png)

前面那里的error可以忽略，因为刚刚那个是还没跑起来就打开网页了
后面的记录就是刚刚对话的记录

**Comment by shenmadouyaowen** [comment]
> > > > > > 我的都不会调用工具 ![微信截图_20231115115525](https://user-images.githubusercontent.com/103934790/283044583-78d185d6-db7f-4390-ad06-57f8886af9cd.png)
> > > > > 
> > > > > 
> > > > > 只有chatglm3-6b才能使用工具
> > > > 
> > > > 
> > > > 我的是chatglm3-6b呀
> > > 
> > > 
> > > 看一下后台终端记录
> > 
> > 
> > ![微信截图_20231115160531](https://user-images.githubusercontent.com/103934790/283045981-c2f7fd0b-3b64-4ada-bf89-749713ae439b.png)
> 
> 前面那里的error可以忽略，因为刚刚那个是还没跑起来就打开网页了 后面的记录就是刚刚对话的记录

去demo_tool.py 文件这行打印看看有没有显示True
![image](https://github.com/THUDM/ChatGLM3/assets/12426724/32063d08-0594-4eab-b744-c4391eee4248)

**Comment by CNUIGB** [comment]
> > > > > > > 我的都不会调用工具 ![微信截图_20231115115525](https://user-images.githubusercontent.com/103934790/283044583-78d185d6-db7f-4390-ad06-57f8886af9cd.png)
> > > > > > 
> > > > > > 
> > > > > > 只有chatglm3-6b才能使用工具
> > > > > 
> > > > > 
> > > > > 我的是chatglm3-6b呀
> > > > 
> > > > 
> > > > 看一下后台终端记录
> > > 
> > > 
> > > ![微信截图_20231115160531](https://user-images.githubusercontent.com/103934790/283045981-c2f7fd0b-3b64-4ada-bf89-749713ae439b.png)
> > 
> > 
> > 前面那里的error可以忽略，因为刚刚那个是还没跑起来就打开网页了 后面的记录就是刚刚对话的记录
> 
> 去demo_tool.py 文件这行打印看看有没有显示True ![image](https://user-images.githubusercontent.com/12426724/283048711-32063d08-0594-4eab-b744-c4391eee4248.png)

![微信截图_20231115163524](https://github.com/THUDM/ChatGLM3/assets/103934790/8f42e82c-fe93-45e1-b717-d033046a6404)
大佬，是FALSE，这是什么原因呀

**Comment by shenmadouyaowen** [comment]
> > > > > > > > 我的都不会调用工具 ![微信截图_20231115115525](https://user-images.githubusercontent.com/103934790/283044583-78d185d6-db7f-4390-ad06-57f8886af9cd.png)
> > > > > > > 
> > > > > > > 
> > > > > > > 只有chatglm3-6b才能使用工具
> > > > > > 
> > > > > > 
> > > > > > 我的是chatglm3-6b呀
> > > > > 
> > > > > 
> > > > > 看一下后台终端记录
> > > > 
> > > > 
> > > > ![微信截图_20231115160531](https://user-images.githubusercontent.com/103934790/283045981-c2f7fd0b-3b64-4ada-bf89-749713ae439b.png)
> > > 
> > > 
> > > 前面那里的error可以忽略，因为刚刚那个是还没跑起来就打开网页了 后面的记录就是刚刚对话的记录
> > 
> > 
> > 去demo_tool.py 文件这行打印看看有没有显示True ![image](https://user-images.githubusercontent.com/12426724/283048711-32063d08-0594-4eab-b744-c4391eee4248.png)
> 
> ![微信截图_20231115163524](https://user-images.githubusercontent.com/103934790/283054117-8f42e82c-fe93-45e1-b717-d033046a6404.png) 大佬，是FALSE，这是什么原因呀

检查一下模型对不对把

**Comment by jasonwongw** [comment]
这是第一次的output
![image](https://github.com/THUDM/ChatGLM3/assets/108513330/31b627b8-a365-48cd-95f8-f9a0bc7b88cf)
这是第二次的output
![image](https://github.com/THUDM/ChatGLM3/assets/108513330/fdd689fb-828e-420d-b431-2e5ca171d425)
因为output的输出少了工具get_weather，之后在处理output时，报错了。
![image](https://github.com/THUDM/ChatGLM3/assets/108513330/758183e1-7941-400b-b3b5-626f79337ab6)

**Comment by GavinZhou0125** [comment]
就是这种情况，第二次输出会不带工具名。考古了下之前的issue说是模型只支持调用一次工具

**Comment by GavinZhou0125** [comment]
我觉得链式调用工具还挺必要的，在垂直场景里问完一个问题要求刷新下页面怎么都有点别扭

**Comment by GavinZhou0125** [comment]
> 这是第一次的output ![image](https://user-images.githubusercontent.com/108513330/283065227-31b627b8-a365-48cd-95f8-f9a0bc7b88cf.png) 这是第二次的output ![image](https://user-images.githubusercontent.com/108513330/283065347-fdd689fb-828e-420d-b431-2e5ca171d425.png) 因为output的输出少了工具get_weather，之后在处理output时，报错了。 ![image](https://user-images.githubusercontent.com/108513330/283065099-758183e1-7941-400b-b3b5-626f79337ab6.png)

感觉模型把这句话识别成单独的对话了？

**Comment by Roych13** [comment]
我估计是要改底层那个“```”的问题，单独提取出calling后面那部分。我抽空研究下。

**Comment by shenmadouyaowen** [comment]
实际上用命令行的时候是没有问题的，应该还是代码哪里判断没有处理好
![image](https://github.com/THUDM/ChatGLM3/assets/12426724/ca9c5838-010d-4946-a757-e26638e72ec9)

**Comment by zRzRzRzRzRzRzR** [comment]
该问题被记录为bad case，相关相似问题若能被复现将会集中一起处理，目前的Function call确实还有哦很多不完善的地方

**Reply by wayfare11** [reply]
这个问题多久处理啊

**Comment by skytodmoon** [comment]
遇到同样得问题，mark

**Comment by zRzRzRzRzRzRzR** [comment]
试一下最新的composite demo

**Comment by fengshaoA** [comment]
如果我在工具定义中使用了多个参数，这个参数该如何定义呢？比如我想查询北京昨天的天气，那么参数中有地点，日期，这种格式的该如何传给模型呢？
目前模型中，好像只接受一个参数的形式，因为 我传的是两个参数，但是到模型中的显示的是一个工具：具体如下：
Action: 
```
{"action": "PassengerFlow", "action_input": "103"}
```<Response [200]>
该问题如何解决呢？@zRzRzRzRzRzRzR

谢谢

**Comment by zRzRzRzRzRzRzR** [comment]
我们langchain tool的yaml或者tool_using 的注册中有提到如何写参数的格式，多参数是支持的

**Comment by greathelin** [comment]
要训练Tool调用，需要先在 ChatGLM3/composite_demo/tool_registry.py 中把tool的函数写上吗？
还是可以先训练，最后部署使用的时候再写上？

**Comment by valkryhx** [comment]
> 要训练Tool调用，需要先在 ChatGLM3/composite_demo/tool_registry.py 中把tool的函数写上吗？ 还是可以先训练，最后部署使用的时候再写上？

现在的版本还支持tool调用能力的微调吗？对比了10.30日的项目内容，新微调demo的代码里面关于tool字段都是not implemented ，感觉不支持tool能力的微调了。

**Reply by huangguifeng** [reply]
不支持了，说是效果不好，给取消了工具微调，不能微调，那这个工具调用有啥用呢。

**Reply by Admin-ppq** [reply]
老版本的代码还能找到嘛？想看看工具微调

**Comment by ATP-BME** [comment]
当需要在多轮对话靠后的轮次进行工具调用时，模型似乎无法成功调用工具。即使给出显式的调用工具的指令也无法成功调用。想请教一下针对这个问题有没有比较好的解决方法

**Reply by ATP-BME** [reply]
尝试修改代码在手动在三轮对话之后再传入一次<system>提示，效果仍然不好

**Comment by dqy1111** [comment]
我想问一下微调后多轮对话为什么只有第一次能输出想要的答案，后面再问就开始胡言乱语了，是数据集的原因还是什么
